File: C/which_side.cpp
#include <utility>
#include "global_variables.h"
#include <math.h>

// returning side, s
std::pair<int,int> which_side(int id_t,int p){
  id_t = id_t - 1;
  p = p - 1;
  double v[2];
  double N[2];
  double scalar;
  double norm_N_1;
  double dot;
  int side;
  int s;
  double ss;
  v[0] = node[p].x - trace_vertex[trace[id_t][0]][0] ;
  v[1] = node[p].y - trace_vertex[trace[id_t][0]][1];
  N[0] = -T[id_t][1];
  N[1] = T[id_t][1];
  dot = v[0]*N[0] + v[1]*N[1];
  norm_N_1 = abs(N[0]) + abs(N[1]);
  scalar = dot/norm_N_1;
  if(scalar < -node[p].toll){
    side = -1;
    s = 0;
  }
  else if(scalar > node[p].toll){
    side = 1;
    s = 0;
  }
  else{
    side = 2;
  }
  ss = ((T[id_t][0]*v[0])+T[id_t][1]*v[1])/T_squared[id_t];
  if (s <= toll_t[id_t] && s >= -toll_t[id_t] ){
    s = 0;
  }
  else{
    s = 1;
  }
  return std::make_pair(side,s);
}

File: 2D/which_side.m
function [side, s] = which_side(id_t,p)
%Per una data traccia e un punto, l'output è:
%side 1 o -1 se non sta sulla retta della traccia 
%side 2 se sta sulla retta della traccia, in questo caso, viena calcolata
%la coordinata curvilinea.

global node
global T
global T_squared
global trace
global trace_vertex
global toll_t

v = [node(p).x - trace_vertex(trace(id_t,1),1);
     node(p).y - trace_vertex(trace(id_t,1),2)];
N = [-T(id_t,2);
     T(id_t,1)];
% scalar rappresenta la distanza punto traccia, per un valore esatto
% dovremmo usare la norma-2, ma è più costosa per via della radice.
scalar = dot(N,v)/norm(N,1);
toll = node(p).toll;
if(scalar<-toll)
    side=-1;
    s=0;
elseif(scalar>toll)
    side=1;
    s=0;
else
    side=2;
    
%conti per trovare la coordinata curvilinea associata.
%In particolare ci serve la norma al quadrato di T. 
%avremo che s = dot(T,v)/(T^2)

    s=dot(T(id_t,:),v)/T_squared(id_t); 
    
    if( s <= toll_t(id_t) && s >= -toll_t(id_t) )
        s = 0;
    elseif (s <= 1 + toll_t(id_t) && s >= 1 - toll_t(id_t))
        s = 1;
    end
end
end


